tasks/push-rpm-manifests-to-pyxis/push-rpm-manifests-to-pyxis.yaml

```diff
@@ -0,0 +1,160 @@
+  steps:
+    - name: download-sbom-files
+      image:
+        quay.io/redhat-appstudio/release-service-utils:b3059c67d4bdb6c9cf1739fdfb2c0cb595fe0f86 #TODO update everywhere
+      volumeMounts:
+        - mountPath: /workdir
+          name: workdir
+      script: |
+        #!/usr/bin/env sh
+        set -eux
+
+        SNAPSHOT_SPEC_FILE="$(workspaces.data.path)/$(params.snapshotPath)"
+        if [ ! -f "${SNAPSHOT_SPEC_FILE}" ] ; then
+            echo "No valid snapshot file was provided."
+            exit 1
+        fi
+
+        IMAGEURLS=($(cat "${SNAPSHOT_SPEC_FILE}" | jq -r '.components[].containerImage'))
+        IMAGEIDS=($(params.containerImageIDs))
+
+        if [ ${#IMAGEURLS[@]} != ${#IMAGEIDS[@]} ]; then
+          echo "ERROR: Lengths of imageURLs and containerImageIDs do not match"
+          exit 1
+        fi
+
+        mkdir /workdir/sboms
+        cd /workdir/sboms
+
+        for i in ${!IMAGEURLS[@]}; do
+          echo "Fetching sbom for image: ${IMAGEURLS[$i]}"
+          cosign download sbom --output-file "${IMAGEIDS[$i]}.json" "${IMAGEURLS[$i]}"
+        done
+
+        SBOM_COUNT=$(ls *.json | wc -l )
+        if [ $SBOM_COUNT != ${#IMAGEURLS[@]} ]; then
+          echo "ERROR: Expected to fetch sbom for ${#IMAGEURLS[@]} images, but only $SBOM_COUNT were saved"
+          exit 1
+        fi
+
+    - name: push-rpm-manifests-to-pyxis
+      image:
+        quay.io/redhat-appstudio/release-service-utils:b3059c67d4bdb6c9cf1739fdfb2c0cb595fe0f86
+      env:
+        - name: pyxisCert
+          valueFrom:
+            secretKeyRef:
+              name: $(params.pyxisSecret)
+              key: cert
+        - name: pyxisKey
+          valueFrom:
+            secretKeyRef:
+              name: $(params.pyxisSecret)
+              key: key
+      volumeMounts:
+        - mountPath: /workdir
+          name: workdir
+      script: |
+        #!/usr/bin/env sh
+        set -eu
+
+        if [[ "$(params.server)" == "production" ]]
+        then
+          export PYXIS_GRAPHQL_API="https://graphql-pyxis.api.redhat.com/graphql/"
+        elif [[ "$(params.server)" == "stage" ]]
+        then
+          export PYXIS_GRAPHQL_API="https://graphql-pyxis.preprod.api.redhat.com/graphql/"
+        else
+          echo "Invalid server parameter. Only 'production' and 'stage' are allowed."
+          exit 1
+        fi
+
+        export PYXIS_CERT_PATH=/workdir/crt
+        export PYXIS_KEY_PATH=/workdir/key
+        echo "${pyxisCert}" > $PYXIS_CERT_PATH
+        echo "${pyxisKey}" > $PYXIS_KEY_PATH
+
+        cd /workdir/sboms
+
+        N=$(params.concurrentLimit)  # The maximum number of images to be processed at once
+        declare -a jobs=()
+        declare -a files=()
+        total=$(ls *.json | wc -l )
+        count=0
+        success=true
+        echo "Starting RPM Manifest upload for $total files in total. " \
+          "Up to $N files will be uploaded at once..."
+
+        for FILE in *.json; do
```

Comment:
Hmm, so this assumes that we have one file per image. But we now create one pyxis image per arch, don't we? https://github.com/konflux-ci/release-service-catalog/blob/development/tasks/create-pyxis-image/create-pyxis-image.yaml#L131

So I think you should use the mapping json that is produced in that task? And do this once for each combination of image and arch?

I actually recall something about hearing from the Pyxis team that we actually don't need to create duplicate pyxis image objects for each arch in pyxis, but I'm not sure where I saw that - having a separate pyxis image per arch would also imply duplicate sbom data being uploaded (but we don't push sboms to pyxis anymore) and I do think that they told us it wouldn't be necessary. But right now I can see we do create those multiple pyxis image objects and since that's the case, I think we should create this for each arch as well.

Reply:
Maybe you are saying L56 is wrong?

Reply:
> shouldn't there be one file per combination of image and arch?

Not really. When you have a multi-arch image, there will still only be one image url (in snapshot spec) and so you will only get one sbom file, but you need to apply this sbom (meaning run the python script) to all the archs for that image - so there will be multiple pyxis images that it's related to. And you will have to look at the json that is produced in create-pyxis-image - `$(workspaces.data.path)/pyxis.json`. 

You can get a hint on what the json will look like from this test: https://github.com/konflux-ci/release-service-catalog/blob/development/tasks/create-pyxis-image/tests/test-create-pyxis-image-one-containerimage-multi-arch.yaml#L111
```sh
              # check if the correct arch and image id are set in the json file
              jq -e '.components[0].pyxisImages[0] | ( .arch == "amd64" ) and ( .imageId == "0001" )' \
                $(workspaces.data.path)/pyxis.json

              jq -e '.components[0].pyxisImages[1] | ( .arch == "ppc64le" ) and ( .imageId == "0002" )' \
                $(workspaces.data.path)/pyxis.json

```

Reply:
Still not getting it. In the line I linked, the arch is different each time. Each one creates a container image ID and adds it to the task result. How is there not one sbom per combination of image and arch?
```
            while IFS= read -r ARCH_DETAIL;
            do
                OS=$(jq -r '.platform.os' <<< "$ARCH_DETAIL")
                ARCH=$(jq -r '.platform.architecture' <<< "$ARCH_DETAIL")
                ARCH_DIGEST=$(jq -r '.digest' <<< "$ARCH_DETAIL")

                <snipped>

                echo $IMAGEID >> $(results.containerImageIDs.path)
                index=$((index + 1))
            done <<< $(get-image-architectures "${PULLSPEC}" | jq -c)
```

Reply:
Yes, the `containerImageIDs` result will contain all the combinations, you're right on that one. The problem is that without the pyxis.json, there is no way to find out which image id belongs to which container image in your task. Because it's not a 1:1 relation. If at least one of your images is multiarch, you will have more image ids than container images and your task will fail with `echo "ERROR: Lengths of imageURLs and containerImageIDs do not match"`. That's why you need the mapping.

Reply:
This is fixed now